source/volume_of_fluid/reconstruct.cc

```diff
@@ -0,0 +1,488 @@
+                  {
+                    typename DoFHandler<dim>::active_cell_iterator curr;
+                    if (cen == endc)
+                      {
+                        curr = endc;
+                      }
+                    else
+                      {
+                        if (j == 0 || j == 2)
+                          {
+                            const unsigned int neighbor_no = 2+(j/2);
+                            const typename DoFHandler<dim>::face_iterator face = cen->face (neighbor_no);
+                            if (face->at_boundary() && !cen->has_periodic_neighbor(neighbor_no)||
+                                face->has_children())
+                              curr = endc;
+                            else
+                              {
+                                const typename DoFHandler<dim>::cell_iterator neighbor =
+                                  cen->neighbor_or_periodic_neighbor(neighbor_no);
+                                if (neighbor->level() == cell->level() &&
+                                    neighbor->active())
+                                  curr = neighbor;
+                                else
+                                  curr = endc;
+                              }
+                          }
+                        else
+                          curr = cen;
+                      }
+                    if (curr != endc)
+                      {
+                        curr->get_dof_indices (cell_dof_indicies);
+                        resc_cell_centers[3 * j + i] = Point<dim> (-1.0 + i,
+                                                                   -1.0 + j);
+                      }
+                    else
+                      {
+                        cell->get_dof_indices (cell_dof_indicies);
+                        resc_cell_centers[3 * j + i] = Point<dim> (0.0,
+                                                                   0.0);
+                      }
+                    local_volume_of_fluids (3 * j + i) = solution (cell_dof_indicies[volume_of_fluid_ind]);
+                    neighbor_cells[3 * j + i] = curr;
+                  }
+              }
+
+            // Gather cell strip sums
+            for (unsigned int i = 0; i < dim * n_sums; ++i)
+              strip_sums[i] = 0.0;
+
+            for (unsigned int i = 0; i < 3; ++i)
+              {
+                for (unsigned int j = 0; j < 3; ++j)
+                  {
+                    strip_sums[3 * 0 + i] += local_volume_of_fluids (3 * j + i);
+                    strip_sums[3 * 1 + j] += local_volume_of_fluids (3 * j + i);
+                  }
+              }
+
+            // std::cout << "  Strip sums: " << std::endl;
+            // for (unsigned int i = 0; i < dim * n_sums; ++i)
+            //   std::cout << "    " << strip_sums[i] << std::endl;
+
+            // Calculate normal vectors for the 6 candidates from the efficient
+            // least squares approach
```

Comment:
This comment needs to be extended significantly and the individual cases in the loop below documented. What are the 6 candidates and why are there 6? How does the efficient least squares approach work (in simple terms)? And what are `di` and `di2`?

Reply:
Added attempted clarification